prov/opx/include/rdma/opx/fi_opx.h

```diff
@@ -0,0 +1,360 @@
+
+static const uint64_t FI_OPX_MEM_TAG_FORMAT		= (0xFFFFFFFFFFFFFFFFULL);
+static const int      FI_OPX_MAX_HFIS				= 16;
+
+/* 
+Users may wish to change the depth of the Rx context ring.
+/sys/module/hfi1/parameters/rcvhdrcnt:8192 is a tuning knob that 
+allows useres to specify the count of entries.
+
+Supported values are 2048 (default), 4096, and 8192
+To save on library bloat, 4096 is not implimented as an optimized compile-time value
+
+Remember that fi_opx_hfi1_poll_once works in terms of 4 doublewords
+So for example, 2k ring length (0-based) mask is
+2047 * 32 = 0xFFE0  
+*/
+static const uint64_t FI_OPX_HDRQ_MASK_RUNTIME	= 0ULL;
+static const uint64_t FI_OPX_HDRQ_MASK_2048 		= 0X000000000000FFE0UL;
+static const uint64_t FI_OPX_HDRQ_MASK_8192		= 0X000000000003FFE0UL;
+
+
+#define FI_OPX_DEFAULT_MSG_ORDER										\
+		(FI_ORDER_ATOMIC_RAR | FI_ORDER_ATOMIC_RAW | FI_ORDER_ATOMIC_WAR | FI_ORDER_ATOMIC_WAW)
+
+#define FI_OPX_TXONLY_CAPS							\
+	( FI_SEND | FI_READ | FI_WRITE | FI_REMOTE_WRITE )
+
+#define FI_OPX_RXONLY_CAPS							\
+	( FI_RECV | FI_DIRECTED_RECV | FI_MULTI_RECV | FI_REMOTE_READ )
+
+#define FI_OPX_BASE_CAPS							\
+	( FI_MSG | FI_TAGGED | OPX_LOCAL_COMM_CAP | OPX_REMOTE_COMM_CAP	\
+	| FI_SOURCE | FI_NAMED_RX_CTX | FI_RMA | FI_ATOMIC )
+
+#define FI_OPX_DEFAULT_CAPS							\
+	(FI_OPX_BASE_CAPS | FI_OPX_TXONLY_CAPS | FI_OPX_RXONLY_CAPS)
+
+#define FI_OPX_DEFAULT_TX_CAPS						\
+	(FI_OPX_BASE_CAPS | FI_OPX_TXONLY_CAPS)
+
+#define FI_OPX_DEFAULT_RX_CAPS						\
+	(FI_OPX_BASE_CAPS | FI_OPX_RXONLY_CAPS)
+
+#define FI_OPX_DEFAULT_MODE							\
+	(FI_CONTEXT2 | FI_ASYNC_IOV)
+
+
+
+#if 0
+/*
+ * Use these when other FI_PROGRESS_* models are implemented.
+ */
+#ifndef OPX_PROGRESS
+#define OPX_PROGRESS FI_PROGRESS_UNSPEC
+#endif
+
+#define IS_PROGRESS_MANUAL(domain_ptr)						\
+	((OPX_PROGRESS == FI_PROGRESS_MANUAL) ||			\
+	((OPX_PROGRESS == FI_PROGRESS_UNSPEC) &&			\
+		((domain_ptr)->data_progress == FI_PROGRESS_MANUAL)))
+
+#define IS_PROGRESS_AUTO(domain_ptr)						\
+	((OPX_PROGRESS == FI_PROGRESS_AUTO) ||			\
+	((OPX_PROGRESS == FI_PROGRESS_UNSPEC) &&			\
+		((domain_ptr)->data_progress == FI_PROGRESS_AUTO)))
+#else
+
+#ifndef OPX_PROGRESS
+#define OPX_PROGRESS FI_PROGRESS_MANUAL
+#endif
+
+#define IS_PROGRESS_MANUAL(domain_ptr)	(1)
+#define IS_PROGRESS_AUTO(domain_ptr)	(0)
+#endif
+
+static inline void always_assert (bool val, char *msg)
+{
+	if (!val) {
+		FI_LOG(fi_opx_global.prov, FI_LOG_DEBUG, FI_LOG_FABRIC,
+				"%s\n", msg);
+		exit(EXIT_FAILURE);
+	}
+}
+
+// Useful for checking that structures are the correct size and other
+// compile-time tests. static_assert has existed since C11 so this
+// should be safe, but we have an else clause just in case.
+#if defined(static_assert)
+#define OPX_COMPILE_TIME_ASSERT(cond, msg) static_assert(cond,msg)
+#else
+#define OPX_COMPILE_TIME_ASSERT(cond, msg) if(0){switch(0){case 0:case cond:;}}
+#endif
+
+static inline void fi_opx_ref_init (int64_t *ref, char *name)
```

Comment:
please avoid 'fi_' prefix.  That is (supposed to be) reserved for application facing API definitions.  This comment applies to a bunch of definitions.  Just use opx_ as the prefix and save your fingers typing.